### PR TITLE
Add all mint related unit tests

### PR DESCRIPTION
**Pull request by ailisp**
Address #146, unit test all corner cases in mint and burn.

### CODE REVIEW

File: environment/commands/transfer-eth-erc20-to-near.js [hunk 1]
```diff
@@ -195,6 +195,9 @@ class TransferETHERC20ToNear {
     )
     // @ts-ignore
     try {
+      console.log('=== proof locker')
```
Comment (nearmax) [inline environment/commands/transfer-eth-erc20-to-near.js:198]: Should we remove this debug code?

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 2]
```diff
@@ -590,6 +601,9 @@ mod tests {
     fn carol() -> AccountId {
         "carol.near".to_string()
     }
+    fn rainbow_bridge_eth_on_near_prover() -> AccountId {
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:604]: What is the purpose of this method?

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 3]
```diff
@@ -27,7 +27,10 @@ Properties inherited from the standard Fungible Token:
 use borsh::{BorshDeserialize, BorshSerialize};
 use near_sdk::collections::{UnorderedMap, UnorderedSet};
 use near_sdk::json_types::U128;
-use near_sdk::{env, ext_contract, near_bindgen, AccountId, Balance, Promise, StorageUsage};
+use near_sdk::{
+    env, ext_contract, near_bindgen, AccountId, Balance, Promise, PromiseOrValue, StorageUsage,
+};
+use serde::Deserialize;
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:33]: What do we need serde for?

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 4]
```diff
@@ -878,4 +892,171 @@ mod tests {
...
+            196, 199, 73, 127, 190, 26, 136, 104, 65, 161, 149, 165, 214, 34, 205, 96, 5, 60, 19,
+            118,
+        ];
+        context.storage_usage = env::storage_usage();
+
+        context.attached_deposit = 1000 * STORAGE_PRICE_PER_BYTE;
+        testing_env!(context.clone());
+
+        contract.burn(
+            1000.into(),
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:912]: Let's make it a local variable

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 5]
```diff
@@ -878,4 +892,171 @@ mod tests {
...
+            "0123456789abcdef0123456789abcdef01234567".into(),
+        );
+        assert_eq!(contract.get_balance(alice()).0, total_supply - 1000);
+        assert_eq!(contract.get_total_supply().0, total_supply - 1000);
+    }
+
+    #[test]
+    #[should_panic(expected = "recipient should be a hex: OddLength")]
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:920]: Nice, thanks for this corner case test. Corner case tests are very important!

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 6]
```diff
@@ -878,4 +892,171 @@ mod tests {
...
+        contract.locker_address = [
+            196, 199, 73, 127, 190, 26, 136, 104, 65, 161, 149, 165, 214, 34, 205, 96, 5, 60, 19,
+            118,
+        ];
+        context.storage_usage = env::storage_usage();
+
+        context.attached_deposit = 1000 * STORAGE_PRICE_PER_BYTE;
+        testing_env!(context.clone());
+
+        let proof: Proof = serde_json::from_reader(
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:978]: I see, why we need serde. We need to leave a comment that serde is only used in tests. We should also make sure that compiler successfully eliminates it during wasm compilation, by making sure that the contract size does not change noticeably after this PR.

File: libs-rs/mintable-fungible-token/src/lib.rs [hunk 7]
```diff
@@ -878,4 +892,171 @@ mod tests {
...
+        contract.mint(proof);
+        assert_eq!(
+            contract.get_balance(rainbow_bridge_eth_on_near_prover()).0,
+            1000
+        );
+        assert_eq!(contract.get_total_supply().0, total_supply + 1000);
+    }
+
+    #[test]
+    #[should_panic(expected = "Event cannot be reused for minting.")]
```
Comment (nearmax) [inline libs-rs/mintable-fungible-token/src/lib.rs:995]: Nice, double-spending corner case.

File: environment/commands/transfer-eth-erc20-to-near.js (hunk 1 above)
Comment (ailisp) [reply environment/commands/transfer-eth-erc20-to-near.js:198]: Yes

File: libs-rs/mintable-fungible-token/src/lib.rs (hunk 2 above)
Comment (ailisp) [reply libs-rs/mintable-fungible-token/src/lib.rs:604]: want to use this account, so follow the pattern of `alice()` `carol()` etc

File: libs-rs/mintable-fungible-token/src/lib.rs (hunk 3 above)
Comment (ailisp) [reply libs-rs/mintable-fungible-token/src/lib.rs:33]: deserialize a proof dumped from eth

File: libs-rs/mintable-fungible-token/src/lib.rs (hunk 6 above)
Comment (ailisp) [reply libs-rs/mintable-fungible-token/src/lib.rs:978]: Yep, I'm thinking of
```
#[not(cfg(test))]
#[derive(..., no SerdeDeserialize)]
struct Proof{
...
}

#[cfg(test)]
#[derive(..., SerdeDeserialize)]
struct Proof{
...
}
```

Do you know a better way to write this?

> We should also make sure that compiler successfully eliminates it during wasm compilation, by making sure that the contract size does not change noticeably after this PR.

Yeah if optimizer works well that's better

**Comment by ailisp** [comment]
Contracts have been moved to rainbow-bridge-rs, so will address comment and recreate PR there
